x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/InternalMapReduceAggregation.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.AggregationReduceContext;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.support.SamplingContext;
+import org.elasticsearch.search.profile.SearchProfileResults;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public final class InternalMapReduceAggregation extends InternalAggregation {
+
+    private final AbstractMapReducer<?, ?, ?> mapReducer;
+    private Writeable mapReduceContext;
+    private final boolean profiling;
+
+    InternalMapReduceAggregation(
+        String name,
+        Map<String, Object> metadata,
+        AbstractMapReducer<?, ?, ?> mapReducer,
+        Writeable mapReduceContext,
```

Comment:
Part of the problem with the casting is probably due to having the three `?`s here.

I think you want the type of `mapReduceContext` to match the type of the second `?`.

So maybe the solution is to make the whole `InternalMapReduceAggregation` a generic type that takes the context type as a parameter.